### DISCUSSION TITLE
UX specs for generalized UTXO models

### DISCUSSION THREAD

**Discussion by Keith-CY** [WalletConnect/walletconnect-monorepo#3344] [Ideas]
Hi team, I'm a developer of Nervos Network(https://github.com/nervosnetwork/) Ecosystem, and our team is exploring a new paradigm to extend wallet connect for generalized UTXO models.

As we know, the UTXO model is definitely different from the account model that a transaction of UTXO model consists of **2 sets of addresses**(**from address set** and **to address set**) while there're only **2 addresses** in a transaction of account model. It should impact the UX protocol dramatically.

For this consideration, we proposed a UX spec of generalized UTXO models for wallet connect protocol:

Here is it: https://github.com/Magickbase/neuron-public-issues/blob/main/Neuron-PRDs/WalletConnect/GeneralComponent.md

In short,
1. The account should be **a set of addresses** specified by the owner, instead of an independent address;
2. Addresses of an account will be **grouped by their algorithms**(functionality/contract/script), so only selected address groups of the account will be accessible to a dapp;
3. A transaction overview will be primarily **displayed with aggregated data** from the perspective of `addresses`, e.g. 

    | address   | amount |
    |-----------|--------|
    | address_a | +10    |
    | address_b | +10    |
    | address_c | -20    |

    while the complex and obscure transaction structure can be inspected manually

    | inputs    | amount | =>  | outputs   | amount |
    | --------- | ------ | --- | --------- | ------ |
    | address_a | 20     | =>  | address_a | 30     |
    |           |        | =>  | address_b | 10     |
    | address_c | 20     | =>  |           |        |

It would be great if a universal UX could be populated by wallet connect, and I appreciate your attention to this matter.

**Comment by bumblefudge** [comment]
Hey @Keith-CY , Juan here from @Chainagnostic . I'm not sure what exactly you're proposing, and whether it's theoretical (a mental model, a design pattern) or practical (a listing in wallet connect's trust network).  If the former, maybe I'm the guy to talk to, if the latter, I'm probably not!

I would mention that the UTXO<>account interoperability model hasn't been battle-tested across many UTXO virtual machines, but it HAS already been integrated into wallet connect's transaction model, and there are already multi-chain dapps and wallets using it (at least with classic BTC forks).  The best description of how this interop is achieved is here:
https://namespaces.chainagnostic.org/bip122/caip10

Hope that's helpful in any case.

**Reply by Keith-CY** [reply]
Hi @bumblefudge 
Appreciate your attention.

We may be mentioning different topics.

https://namespaces.chainagnostic.org/bip122/caip10 is an addition specifically to `account spec` at https://github.com/ChainAgnostic/CAIPs/blob/main/CAIPs/caip-10.md, it defines an account for Bitcoin/UTXO by following BIP13, and it's about `namespace` protocol.

We're also defining the addition to `account spec` for nervos network at https://github.com/Magickbase/neuron-public-issues/issues/148

However, the topic I want to start is the User Experience (UX) of UTXO model on the wallet side.

During the work of `namespace`, we found the well-defined component [web3modal](https://web3modal.com/) has become the de facto standard on the dapp side for unified UX, but there's a missing counterpart on the wallet side.

The interaction of account model is relatively simple, and it may not have received much attention on the UX as a result. But UTXO model is much more complex because its transaction structure is flexible and composable.

So the idea comes here: we would like to define a well-designed UTXO interaction on the wallet side,  deliver a general component similar to [web3modal](https://web3modal.com/) for onboarding, and try to populate it among the wallets using wallet connect for a unified UX.

**Reply by bumblefudge** [reply]
Ah, OK, that's very interesting.

So, I would propose the following two parallel processes:
1.) See who wants to work on this.  If there was significant buy-in from any WC-registered wallets that this kind of UX component would be used, and it fit in well with the WC SDKs, it could be included in https://github.com/WalletConnect/awesome-walletconnect and otherwise recommended to WC's community of wallets. If it works for all of the BIP122 family of chains and forks, or better yet, if you can confirm your design works with other kinds of UTXO account systems that use different keys or primitives, it should be pretty easy to find an interest wallet or two to prototype this with! I could also reach out to the Metamask Snaps team to see if anyone is working on UTXO<>MM UX.
2.) In parallel, I would recommend an ALTERNATIVE to CAIP-10, i.e. a new CAIP, that describes as generally as  possible this multi-UTXO approach (would this be like different UTXO for different pay scripts/functions?).  Then, a BIP122-specific profile of that which enumerates the key types, address format, names of the pay scripts to date, etc.  It could be the same people doing the prototyping, or others, but it's additional work that improves the chances of this being adopted by more wallets.

and, optionally,

3.) Consider doing a lightning talk at the wallet unconference in Istanbul in November! The audience will contain many many wallet developers, so it would be perfect for adoption if you've got 1 and 2 to a stable draft by then 😎
